tests/sync_filter_test.go

```diff
@@ -14,23 +18,54 @@ func TestSyncFilter(t *testing.T) {
 	defer deployment.Destroy(t)
 	authedClient := deployment.Client(t, "hs1", "@alice:hs1")
 	// sytest: Can create filter
-	reqBody, err := json.Marshal(map[string]interface{}{
-		"room": map[string]interface{}{
-			"timeline": map[string]int{
-				"limit": 10,
+	t.Run("Can create filter", func(t *testing.T) {
+		reqBody, err := json.Marshal(map[string]interface{}{
+			"room": map[string]interface{}{
+				"timeline": map[string]int{
+					"limit": 10,
+				},
 			},
-		},
+		})
+		if err != nil {
+			t.Fatalf("failed to marshal JSON request body: %s", err)
+		}
+		createFilter(authedClient, reqBody, t, "@alice:hs1")
 	})
-	if err != nil {
-		t.Fatalf("failed to marshal JSON request body: %s", err)
-	}
-	t.Run("Can create filter", func(t *testing.T) {
-		res := authedClient.MustDo(t, "POST", []string{"_matrix", "client", "r0", "user", "@alice:hs1", "filter"}, json.RawMessage(reqBody))
-
+	t.Run("Can download filter", func(t *testing.T) {
+		reqBody, err := json.Marshal(map[string]interface{}{
+			"room": map[string]interface{}{
+				"timeline": map[string]int{
+					"limit": 10,
+				},
+			},
+		})
+		if err != nil {
+			t.Fatalf("failed to marshal JSON request body: %s", err)
+		}
+		filterID := createFilter(authedClient, reqBody, t, "@alice:hs1")
+		res := authedClient.MustDo(t, "GET", []string{"_matrix", "client", "r0", "user", "@alice:hs1", "filter", filterID}, nil)
 		must.MatchResponse(t, res, match.HTTPResponse{
 			JSON: []match.JSON{
-				match.JSONKeyPresent("filter_id"),
+				match.JSONKeyPresent("room"),
+				match.JSONKeyEqual("room.timeline.limit", "10"),
```

Comment:
@Kegsay I am getting error
```sh
=== CONT  TestSyncFilter/Can_download_filter
    sync_filter_test.go:47: MatchResponse key 'room.timeline.limit' got '10' want '10' - http://localhost:55093/_matrix/client/r0/user/@alice:hs1/filter/1 => {"event_format":"client","presence":{"limit":20},"account_data":{"limit":20},"room":{"ephemeral":{"limit":20},"state":{"limit":2147483647},"timeline":{"limit":10},"account_data":{"limit":20}}}
```,
I even tried to remove 10 from the double quotes but still the error persisted. Any idea where I am going wrong?

Reply:
Interesting point. So when Go deserialises bytes into JSON, it converts all numbers into `float64`:

> float64, for JSON numbers

https://golang.org/pkg/encoding/json/#Unmarshal

When you use `JSONKeyEqual` it will type-check and value-check the value, in this case `10`. Referring to a number directly in Go code defaults to type `int` as per the [Go specification](https://golang.org/ref/spec#Integer_literals). This is why removing the double quotes doesn't work. Instead, type-cast `10` to float64 e.g `float64(10)`.

Reply:
Thanks! Got it!